doc/developer/design/20320318_persist_columnar.md

```diff
@@ -0,0 +1,608 @@
+
+<table>
+<tr>
+<td> Scalar Type </td><td> Rust Representation </td><td> Arrow Array </td><td> Notes </td>
+</tr>
+
+<tr>
+<td> Numeric </td>
+<td> 
+
+```rust
+struct Decimal<const N: usize> {
+  digits: u32,
+  exponent: i32,
+  bits: u8,
+  lsu: [u16; N],
+}
+``` 
+
+</td>
+<td>
+
+```
+StructArray<{
+  lossy: f64,
+  actual: Bytes,
+}>
+```
+
+</td>
+<td>
+
+Encoded as two values, a lossy `f64` for sorting and filtering, then a
+serialized representation of `Decimal` as opaque bytes.
+
+See <https://speleotrove.com/decimal/dnnumb.html> for an explanation as to what
+the fields in the Rust type represent.
+
+</td>
+
+<tr>
+<td> Date </td>
+<td> 
+
+```rust
+struct Date {
+  days: i32,
+}
+``` 
+
+</td>
+<td>
+
+`PrimitiveArray<i32>`
+
+</td>
+<td>
+
+Directly encode the number of days.
+
+</td>
+
+<tr>
+<td> Time </td>
+<td> 
+
+```rust
+struct NaiveTime {
+  secs: u32,
+  frac: u32,
+}
+``` 
+
+</td>
+<td>
+
+`PrimitiveArray<u64>`
+
+</td>
+<td>
+
+Represented as the number of microseconds since midnight.
+
+> **Alternative:** Instead of representing this as the number of microseconds
+since midnight we could stich together the two `u32`s from `NaiveTime`. This
+would maybe save some CPU cycles during encoding and decoding, but probably not
+enough to matter, and it locks us into a somewhat less flexible format.
+
+> Note: We only need 37 bits to represent this total range, leaving 27 bits 
+unused. In the future if we support the `TIMETZ` type we could probably also
+represent that in a `u64`, using these extra bits to store the timezone.
+
+</td>
+
+<tr>
+<td> Timestamp </td>
+<td> 
+
+```rust
+struct NaiveDateTime {
+  date: NaiveDate {
+    // (year << 13) | day
+    ymdf: i32,
+  },
+  time: NaiveTime,
+}
+``` 
+
+</td>
+<td>
+
+`PrimitiveArray<i64>`
+
+</td>
+<td>
+
+`chrono` (our underlying date time library) uses a more memory efficient
+encoding of date by squeezing both year and day into a single `i32`, combined
+with a `NaiveTime` this ends up being 12 bytes. 
+
+
+We can repesent this same range of time as the number of microseconds since 
+the UNIX epoch in an `i64`. Postgres does something very similar, the only 
+difference is it uses an offset of 2000-01-01.
+
+</td>
+
+<tr>
+<td> TimestampTz </td>
+<td> 
+
+```rust
+struct DateTime<Tz: TimeZone> {
+    datetime: NaiveDateTime,
+    // purely type info
+    offset: Tz::Offset,
+}
+``` 
+
+</td>
+<td>
+
+`PrimitiveArray<i64>`
+
+</td>
+<td>
+
+Just like Timestamp, we'll encode this as the number of microseconds since
+the UNIX epoch. We don't actually need to store any timezone information, 
+instead we convert to the session timezone when loaded. This is how both 
+Postgres works.
+
+</td>
+
+<tr>
+<td> Interval </td>
+<td> 
+
+```rust
+struct Interval {
+  months: i32,
+  days: i32,
+  micros: i64,
+}
+``` 
+
+</td>
+<td>
+
+`FixedSizeBinary[16]`
+
+</td>
+<td>
+
+Represented by encoding the `months`, `days`, and `micros` fields encoded as 
+little endian. 
```

Comment:
How do these sort in postgres? We might be able to get meaningful sorting if we encode it in big endian instead (right? isn't that how that works?)

Reply:
You're totally right, if we want them to sort by `month, day, micros` then big endian is how we would want to encode these, updated!

I'll double check they sort in Postgres

Reply:
Checked on this, and if we encode the values as big endian, and apply an offset for negative numbers, we do retain the same order. Here's a draft PR https://github.com/MaterializeInc/materialize/pull/26799

Also should note that Intervals do not currently sort correctly in Materialize, https://github.com/MaterializeInc/materialize/issues/10854. I don't think this matters for this project, but is worth calling out